lib/spidr/agent/sitemap.rb

```diff
@@ -0,0 +1,69 @@
+require 'set'
+
+module Spidr
+  class Agent
+    # Common locations for Sitemap(s)
+    COMMON_SITEMAP_LOCATIONS = %w[
+      sitemap.xml
+      sitemap.xml.gz
+      sitemap.gz
+      sitemap_index.xml
+      sitemap-index.xml
+      sitemap_index.xml.gz
+      sitemap-index.xml.gz
+    ].freeze
+
+    #
+    # Initializes the sitemap fetcher.
+    #
+    def initialize_sitemap
+      @sitemap = true
+    end
+
+    #
+    # Returns the URLs found as per the sitemap.xml spec.
+    #
+    # @return [Array<URI::HTTP>, Array<URI::HTTPS>]
+    #   The URLs found.
+    #
+    # @see https://www.sitemaps.org/protocol.html
+    def sitemap_urls(url)
+      return [] unless @sitemap
+      base_url = to_base_url(url)
+
+      if @robots
+        if urls = @robots.other_values(base_url)['Sitemap']
+          return urls.flat_map { |u| get_sitemap_urls(url: u) }
+        end
+      end
```

Comment:
This is a clever way of populating the the queue using the `Sitemap` listed in the `robots.txt` file. Although, I feel like we should not request any URL before `run` has been called. A way around that would be to add `every_robots_txt` and `every_sitemap_xml` callback hooks, and use those to automatically parse and enqueue URLs when those files are encountered.

Something like:

```ruby
  agent.every_robots_txt do |robots| # RobotsTXT class
    # check robots for a `Sitemap` entry
  end
  
  agent.every_sitemap_xml do |sitemap| # SitemapXML class
    sitemap.urls.each { |url| agent.enqueue(url) }
  end
```

Reply:
Hmm, `every_robots_txt` might not current since I forgot the Robots library automatically fetches `/robots.txt` files and caches them when you query it. I could possibly add a `every_host` callback to detect when the Agent visits a new hostname, and then we can use that to eager request `/robots.txt` and `/sitemap.xml`. This would also allow the sitemap detection logic to fire on every new host/domain we spider, not just the first one. Thoughts?

I could add a `every_host` callback hook in the 0.7.0 branch, if you think it's a good idea.